From 374f2d2a4f8c51a781627610e3d5afd9b7e1fa74 Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Tue, 7 Feb 2017 02:03:32 +0800
Subject: [PATCH 0797/1345] clk: a3700: add GBE BM clock support

commit  17b1e8f21fb357b962f328e83594354202d0701d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Added the support of GBE core clock rate get and clock gating.

Change-Id: I91e2a565d3c49717d593ca93a30564eabed5743a
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36454
Reviewed-by: Yelena Krivosheev <yelena@marvell.com>
Verified-Armada37x0: Yelena Krivosheev <yelena@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/mvebu/armada-3700.c |   15 +++++++++++++++
 drivers/clk/mvebu/armada-3700.h |    1 +
 2 files changed, 16 insertions(+)

diff --git a/drivers/clk/mvebu/armada-3700.c b/drivers/clk/mvebu/armada-3700.c
index e7180de..f454138 100644
--- a/drivers/clk/mvebu/armada-3700.c
+++ b/drivers/clk/mvebu/armada-3700.c
@@ -110,6 +110,7 @@ enum {
 	A3700_TBG_TO_GBE0_CLK,
 	A3700_TBG_TO_GBE1_CLK,
 	A3700_TBG_TO_SPI_CLK,
+	A3700_TBG_TO_GBE_BM_CLK,
 };
 
 static const struct coreclk_ratio armada_3700_coreclk_ratios[] __initconst = {
@@ -121,6 +122,7 @@ enum {
 	{ .id = A3700_TBG_TO_GBE0_CLK, .name = "gbe0-core" },
 	{ .id = A3700_TBG_TO_GBE1_CLK, .name = "gbe1-core" },
 	{ .id = A3700_TBG_TO_SPI_CLK, .name = "spi-core" },
+	{ .id = A3700_TBG_TO_GBE_BM_CLK, .name = "gbe-bm-core" },
 };
 
 /***************************************************************************************************
@@ -208,6 +210,18 @@ static void __init armada_3700_get_clk_ratio(
 		*div = prscl1 * prscl2;
 		break;
 
+	case A3700_TBG_TO_GBE_BM_CLK:
+		*tbg = (clkr32(MVEBU_SOUTH_CLOCK_TBG_SELECT_REG) >> TBG_GBE_CORE_CLK_SEL_OFFSET) &
+					MVEBU_TBG_CLK_SEL_MASK;
+		prscl1 = (clkr32(MVEBU_SOUTH_CLOCK_DIVIDER_SELECT1_REG) >> GBE_CORE_CLK_PRSCL1_OFFSET) &
+					MVEBU_TBG_CLK_PRSCL_MASK;
+		prscl2 = (clkr32(MVEBU_SOUTH_CLOCK_DIVIDER_SELECT1_REG) >> GBE_CORE_CLK_PRSCL2_OFFSET) &
+					MVEBU_TBG_CLK_PRSCL_MASK;
+		div2 = (clkr32(MVEBU_SOUTH_CLOCK_DIVIDER_SELECT1_REG) >> GBE_BM_CORE_CLK_DIV_OFFSET) &
+					MVEBU_TBG_CLK_DIV_MASK;
+		*div = (prscl1 * prscl2) << div2;
+		break;
+
 	default:
 		*div = 0;
 		break;
@@ -319,6 +333,7 @@ static void __init armada_3700_coreclk_init(struct device_node *np)
 static const struct clk_gating_soc_desc armada_3700_south_bridge_gating_desc[] __initconst = {
 	{ "gbe1-gate", "gbe1-core", 4, 0, CLK_GATE_SET_TO_DISABLE },
 	{ "gbe0-gate", "gbe0-core", 5, 0, CLK_GATE_SET_TO_DISABLE },
+	{ "gbe-bm-gate", "gbe-bm-core", 9, 0, CLK_GATE_SET_TO_DISABLE },
 	{ "pcie", NULL, 14, 0, CLK_GATE_SET_TO_DISABLE },
 	{ "usb32-ss-sys-gate", "usb32-ss-sys", 17, 0, CLK_GATE_SET_TO_DISABLE },
 	{ }
diff --git a/drivers/clk/mvebu/armada-3700.h b/drivers/clk/mvebu/armada-3700.h
index a9da0eb..0f13e42 100644
--- a/drivers/clk/mvebu/armada-3700.h
+++ b/drivers/clk/mvebu/armada-3700.h
@@ -130,6 +130,7 @@ struct a3700_clk_desc {
 #define GBE_CORE_CLK_PRSCL2_OFFSET		(18)
 #define GBE0_CORE_CLK_DIV_OFFSET		(14)
 #define GBE1_CORE_CLK_DIV_OFFSET		(13)
+#define GBE_BM_CORE_CLK_DIV_OFFSET		(12)
 
 #define MVEBU_SOUTH_CLOCK_DIVIDER_SELECT2_REG	(MVEBU_SOUTH_CLOCK_REGS_BASE + 0xC)
 
-- 
1.7.9.5

